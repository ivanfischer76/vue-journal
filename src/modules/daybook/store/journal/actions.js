import journalApi from '@/api/journalApi'

export const createEntry = async({commit}, entry) => {
	const { data } = await journalApi.post(`/entries.json`, entry)
	entry.id = data.name
	commit('addEntry', { ...entry }) 
	return entry.id
}

export const deleteEntry = async({commit}, id) => {
	const resp = await journalApi.delete(`/entries/${id}.json`)
	commit('deleteEntry', id)
	return id
}

export const loadEntries = async ({commit}) => {
	const { data } = await journalApi.get('/entries.json')
	if(!data){
		commit('setEntries', [])
		return
	}
	const entries = []
	for(let id of Object.keys(data)){
		entries.push({
			id,
			...data[id]
		})
	}
	commit('setEntries', entries)
}

export const updateEntry = async({commit}, entry) => {
	const { date, picture, text } = entry
	const dataToSave = { date, picture, text }
	const resp = await journalApi.put(`/entries/${entry.id}.json`, dataToSave)
	dataToSave.id = entry.id
	commit('updateEntry', { ...dataToSave }) 
	// { ...dataToSave } es para que no lo pase por referencia porque sino al modificaro 
	// intentarÃ­a modificar el state directamente y no se permite
}


// export const myAction = async({commit}) => {

// }