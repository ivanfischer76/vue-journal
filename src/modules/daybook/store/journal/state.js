

export default () => ({
	isLoading: true,
	entries: [
		{
			id: new Date().getTime(),
			date: new Date().toDateString(),
			text: 'Nisi sunt minim eiusmod in. Adipisicing quis incididunt amet voluptate anim culpa consectetur est. Tempor eu elit dolore qui proident. Anim cupidatat cillum tempor et culpa laboris. Dolor aute aliqua cupidatat eu nostrud.',
			picture: null,
		},
		{
			id: new Date().getTime() + 1000,
			date: new Date().toDateString(),
			text: 'Aliqua dolore excepteur id aliqua ad fugiat et commodo Lorem sunt ea non est excepteur. Reprehenderit culpa sint mollit nulla ex adipisicing ullamco cillum voluptate. Occaecat minim ea est sunt consectetur officia nisi amet ex nisi dolore enim. Quis esse proident in commodo ut culpa tempor cupidatat exercitation qui minim labore elit. Reprehenderit et occaecat sint Lorem id proident commodo nostrud adipisicing.',
			picture: null,
		},
		{
			id: new Date().getTime() + 2000,
			date: new Date().toDateString(),
			text: 'Aute deserunt veniam in cillum nostrud nulla dolore id. Ipsum aliquip veniam reprehenderit duis consequat id exercitation qui minim. Fugiat est eu fugiat in ullamco culpa ullamco ut occaecat do. Amet aliqua eu duis incididunt cupidatat commodo. Ipsum ea et consectetur amet est proident ullamco magna laborum do et sint esse ad. Aute voluptate sint voluptate eiusmod dolore adipisicing dolore eu aliqua dolore ex excepteur et.',
			picture: null,
		},
	]
})