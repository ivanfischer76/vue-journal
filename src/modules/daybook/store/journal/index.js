
import * as actions from './actions'
import * as getters from './getters'
import * as mutations from './mutations'
import state from './state'

const journal = {
	namespaced: true,
	actions,
	getters, 
	mutations,
	state
}

export default journal