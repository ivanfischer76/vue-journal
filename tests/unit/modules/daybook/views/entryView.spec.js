import { createStore } from "vuex"
import { shallowMount } from "@vue/test-utils"
import journal from '@/modules/daybook/store/journal'
import { journalState } from '../../../mock-data/test-journal-state'
import EntryView from '@/modules/daybook/views/EntryView'
import Swal from  'sweetalert2'

const createVuexStore = ( initialState ) => createStore({
	modules: {
		journal: {
			...journal,
			state: { ...initialState}
		}
	}
})

jest.mock('sweetalert2', () => ({
	fire: jest.fn(),
	showLoading: jest.fn(),
	close: jest.fn()
}))

describe('Pruebas en el EntryView', () => {
	const store = createVuexStore ( journalState )
	store.dispatch = jest.fn() //evita que se disparen los dispatch y borren cosas
	const mockRouter = {
		push: jest.fn()
	}
	let wrapper 
	beforeEach(() => {
		jest.clearAllMocks()
		wrapper = shallowMount( EntryView, {
			props: {
				id: '-MkrF2ex-rNJra06xK24'
			},
			global: {
				mocks: {
					$router: mockRouter
				},
				plugins: [ store ]
			}
		})
	})

	test('debe sacar al usuario porque el id no existe', () => {
		const wrapper = shallowMount( EntryView, {
			props: {
				id: 'Este ID no existe en el store'
			},
			global: {
				mocks: {
					$router: mockRouter
				},
				plugins: [ store ],
			}
		})
		expect(mockRouter.push).toHaveBeenCalledWith({ name: 'no-entry' })
	})

	test('debe mostrar la entrada correctamente', () => {
		expect(wrapper.html()).toMatchSnapshot()
		expect(mockRouter.push).not.toHaveBeenCalled()
	})

	test('debe borrar la entrada y salir', (done) => {
		Swal.fire.mockReturnValueOnce(Promise.resolve({ isConfirmed: true }))
		// estímulo
		wrapper.find('.btn-danger').trigger('click')
		// respuesta esperada	
		expect(Swal.fire).toHaveBeenCalledWith({
			title: '¿Está seguro?',
			text: 'Una ve borrado, no se puede recuperar',
			showDenyButton: true,
			confirmButtonText: 'Si, estoy seguro'
		})
		//el setTimeout resuelve un problema de tiempo de respuesta
		setTimeout( () => {
			expect(store.dispatch).toHaveBeenCalledWith('journal/deleteEntry', '-MkrF2ex-rNJra06xK24')
			expect(mockRouter.push).toHaveBeenCalled()
			done()
		}, 1000)
	})
})