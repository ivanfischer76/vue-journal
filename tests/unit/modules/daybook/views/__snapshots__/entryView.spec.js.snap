// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el EntryView debe mostrar la entrada correctamente 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="mx-1 fs-5">Jueves,</span><span class="text-success fs-5 fw-bold">30</span><span class="mx-1 fs-5"> de Septiembre de</span><span class="mx-1 fs-5">2021</span></div>
  <div><input type="file" accept="image/png, image/jpeg" style="display: none;"><button class="btn btn-danger mx-2"> Borrar <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary"> Subir foto <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="¿Que sucedió hoy?"></textarea></div>
<fab-stub icon="fa-save"></fab-stub><img src="https://res.cloudinary.com/dbmvt3wlh/image/upload/v1633014673/daybook/wveyzwfxqslc0ntnzdtc.jpg" alt="entry-picture" class="img-thumbnail">
<!--v-if-->
`;
